File: lib/pdf/encolher.ts
'use client';

/**
 * Fazer um PDF caber num tamanho, no navegador.
 *
 * O irmão deste arquivo em Python (`motor/motor/encolher.py`) faz melhor: ele
 * reduz **só as imagens embutidas** e o texto continua texto. Aqui isso não
 * existe, e vale dizer por quê em vez de deixar a diferença por conta do
 * acaso: nem o pdf-lib nem o pdf.js sabem reescrever uma imagem dentro de um
 * PDF. O pdf-lib monta arquivos, o pdf.js desenha páginas — e desenhar a
 * página inteira é o único caminho que o navegador oferece.
 *
 * Então, no site, a página vira foto. É perda de verdade: o texto deixa de
 * ser selecionável e a cor passa pelo sRGB do canvas. A nota diz isso com
 * todas as letras, e no aplicativo o caminho é o outro.
 *
 * A ordem das tentativas é a mesma dos dois lados, e vai do que não custa
 * nada para o que custa: primeiro reescrever a estrutura, e só depois mexer
 * no desenho.
 */

import { canvasToBlob, openWithPdfJs, openWithPdfLib, renderPageToCanvas, respirar, salvarPdf } from './nucleo';
import { loadPdfLib } from './lazy';
import type { RunContext } from './tipos';

/**
 * A escada de tentativas.
 *
 * 200 DPI ainda imprime bem em laser; 150 é leitura de tela confortável; 96 é
 * onde texto dentro de imagem ainda se lê. Abaixo disso já é o caso de
 * "precisa caber, custe o que custar".
 */
export const DEGRAUS = [
  { dpi: 200, qualidade: 0.88 },
  { dpi: 150, qualidade: 0.82 },
  { dpi: 120, qualidade: 0.75 },
  { dpi: 96, qualidade: 0.68 },
  { dpi: 72, qualidade: 0.6 },
  { dpi: 50, qualidade: 0.45 },
];

export type Encolhido = {
  blob: Blob;
  coube: boolean;
  /** Falso quando alguma página teve que ser redesenhada. */
  semPerda: boolean;
};

/** Redesenha o documento inteiro no DPI e na qualidade pedidos. */
async function rasterizar(
  bytes: ArrayBuffer,
  senha: string | undefined,
  dpi: number,
  qualidade: number,
  ctx: RunContext,
  canvas: HTMLCanvasElement,
): Promise<Blob> {
  const { PDFDocument } = await loadPdfLib();
  const doc = await openWithPdfJs(bytes, senha);
  const out = await PDFDocument.create();

  try {
    for (let i = 1; i <= doc.numPages; i += 1) {
      const page = await doc.getPage(i);
      const { widthPt, heightPt } = await renderPageToCanvas(page, dpi, canvas);
      const jpeg = await canvasToBlob(canvas, 'image/jpeg', qualidade);
      const embutida = await out.embedJpg(await jpeg.arrayBuffer());
      out.addPage([widthPt, heightPt]).drawImage(embutida, { x: 0, y: 0, width: widthPt, height: heightPt });
      page.cleanup();
      await respirar(ctx);
    }
  } finally {
    await doc.destroy();
  }

  return salvarPdf(out, senha);
}

/**
 * Encolhe até caber, tirando o menos possível.
 *
 * Devolve o menor arquivo que conseguiu mesmo quando não coube: entregar
 * nada seria pior, e quem chamou avisa que passou.
 */
export async function ateCaber(
  bytes: ArrayBuffer,
  limiteBytes: number,
  ctx: RunContext,
  senha?: string,
): Promise<Encolhido> {
  // Passo 1: só reescrever a estrutura. Não custa qualidade nenhuma.
  const arrumado = await salvarPdf(await openWithPdfLib(bytes, senha), senha);
  if (arrumado.size <= limiteBytes) {
    return { blob: arrumado, coube: true, semPerda: true };
  }

  const canvas = document.createElement('canvas');
  let melhor = arrumado;

  for (const degrau of DEGRAUS) {
    const tentativa = await rasterizar(bytes, senha, degrau.dpi, degrau.qualidade, ctx, canvas);
    melhor = tentativa;
    if (tentativa.size <= limiteBytes) {
      return { blob: tentativa, coube: true, semPerda: false };
    }
  }

  // Nem no degrau mais baixo. Se depois de tudo o redesenho ficou maior que a
  // simples reescrita — acontece em página só de texto —, o menor dos dois é
  // o que vale.
  const escolhido = melhor.size < arrumado.size ? melhor : arrumado;
  return { blob: escolhido, coube: false, semPerda: escolhido === arrumado };
}

File: motor/motor/encolher.py
"""Fazer um PDF caber num tamanho, tirando o menos possivel.

Nasceu de um caso concreto: dividindo um documento com limite de 1 MB por
parte, algumas partes saiam com 3,9 MB. Nao era defeito da divisao — uma
pagina sozinha nao tem como ser dividida de novo. Para caber, ela precisa
encolher.

A ordem das tentativas e o que importa aqui, e ela vai do que nao custa nada
para o que custa:

1. **Arrumar a estrutura.** Deduplicar objetos repetidos, jogar fora o que
   ninguem referencia, recomprimir os fluxos. Nao mexe em um pixel sequer: o
   resultado e identico ao original, so que menor. Muitas vezes ja resolve.

2. **Reduzir as imagens embutidas.** So as imagens; o texto continua texto,
   selecionavel e pesquisavel. E o que o pessoal chama de "comprimir PDF sem
   estragar" e o que a maioria dos servicos online faz.

O que este arquivo **nao** faz e redesenhar a pagina inteira como foto. Isso
corta muito mais, mas destroi o texto — e virou o padrao de tanta ferramenta
que a pessoa acha normal receber de volta um documento que nao da mais para
copiar. Aqui isso e outra ferramenta, pedida de proposito.
"""

from __future__ import annotations

import os
from typing import Any, Dict, List, Optional

import pymupdf

from .documento import salvar

# A escada de tentativas, da mais conservadora para a mais agressiva.
#
# O `rewrite_images` do MuPDF reduz por metades — uma imagem de 2480 px vai
# para 1240 e depois para 620 —, entao degraus vizinhos as vezes dao o mesmo
# arquivo. Nao ha desperdicio real: a primeira que couber encerra a busca, e
# medir e barato perto de gravar.
#
# 200 DPI ainda imprime bem em laser; 150 e leitura de tela confortavel; 96 e
# o limite em que texto dentro de imagem ainda se le; abaixo disso e para o
# caso de "precisa caber, custe o que custar".
DEGRAUS = [
    {"dpi": 200, "qualidade": 88, "conta": "imagens acima de 200 DPI reduzidas"},
    {"dpi": 150, "qualidade": 82, "conta": "imagens reduzidas para 150 DPI"},
    {"dpi": 120, "qualidade": 75, "conta": "imagens reduzidas para 120 DPI"},
    {"dpi": 96, "qualidade": 68, "conta": "imagens reduzidas para 96 DPI"},
    {"dpi": 72, "qualidade": 60, "conta": "imagens reduzidas para 72 DPI, o limite do que ainda se le"},
    {"dpi": 50, "qualidade": 45, "conta": "imagens reduzidas ao minimo"},
]


class Resultado:
    """O que aconteceu com um arquivo, para a nota poder contar a verdade."""

    def __init__(self, bytes_saida: int, coube: bool, conta: str, sem_perda: bool):
        self.bytes = bytes_saida
        self.coube = coube
        self.conta = conta
        self.sem_perda = sem_perda


def apenas_arrumando(
    origem: str,
    destino: str,
    senha: str = "",
) -> int:
    """Grava de novo sem tocar em imagem nenhuma. Devolve o tamanho.

    E o passo que nunca custa qualidade. `salvar` ja liga a deduplicacao e a
    recompressao, entao aqui basta abrir e gravar.

    O arquivo e lido para a memoria antes de abrir, e nao aberto pelo caminho.
    A diferenca aparece quando `destino` e o proprio `origem` — que e como o
    dividir chama, encolhendo a parte por cima dela mesma: aberto pelo
    caminho, o MuPDF recusa gravar no arquivo de onde esta lendo.
    """
    with open(origem, "rb") as arquivo:
        conteudo = arquivo.read()

    doc = pymupdf.open(stream=conteudo, filetype="pdf")
    try:
        if senha:
            doc.authenticate(senha)
        return salvar(doc, destino, senha)
    finally:
        doc.close()


def ate_caber(
    origem: str,
    destino: str,
    limite_bytes: int,
    senha: str = "",
    degraus: Optional[List[Dict[str, Any]]] = None,
    andamento=None,
) -> Resultado:
    """Encolhe ate caber no limite, tirando o menos possivel.

    `origem` e `destino` podem ser o mesmo caminho: cada tentativa reabre o
    arquivo de origem, entao o destino so e escrito no fim de cada rodada.
    Isso importa porque o `rewrite_images` altera o documento no lugar, e
    tentar o degrau seguinte em cima do anterior encolheria duas vezes.
    """
    degraus = DEGRAUS if degraus is None else degraus

    # Passo 1: so arrumar. Sem perder nada.
    tamanho = apenas_arrumando(origem, destino, senha)
    if tamanho <= limite_bytes:
        return Resultado(tamanho, True, "arrumando a estrutura, sem tocar nas imagens", True)

    # Guarda o original: as tentativas seguintes todas partem dele, e o
    # destino pode ser o proprio arquivo de origem.
    with open(origem, "rb") as arquivo:
        original = arquivo.read()

    melhor = tamanho
    melhor_conta = "arrumando a estrutura, sem tocar nas imagens"

    for indice, degrau in enumerate(degraus):
        if andamento:
            andamento(indice / len(degraus), f"Tentando caber em {limite_bytes // 1024} KB")

        doc = pymupdf.open(stream=original, filetype="pdf")
        try:
            if senha:
                doc.authenticate(senha)
            # `dpi_threshold` tem que ser maior que `dpi_target`: ele diz a
            # partir de qual resolucao vale mexer na imagem.
            doc.rewrite_images(
                dpi_threshold=degrau["dpi"] + 1,
                dpi_target=degrau["dpi"],
                quality=degrau["qualidade"],
            )
            tamanho = salvar(doc, destino, senha)
        finally:
            doc.close()

        melhor = tamanho
        melhor_conta = degrau["conta"]
        if tamanho <= limite_bytes:
            return Resultado(tamanho, True, degrau["conta"], False)

    # Nao coube nem no degrau mais baixo. O arquivo gravado e o menor que se
    # conseguiu, e quem chamou avisa que passou.
    return Resultado(melhor, False, melhor_conta, False)
